test/model/typing_status_test.dart

```diff
@@ -0,0 +1,278 @@
+  final groupNarrow = DmNarrow.withOtherUsers(
+    [otherUser.userId, thirdUser.userId], selfUserId: eg.selfUser.userId);
+
+  anyTimer(Subject<Object?> it) => it.isA<Timer>();
+
+  group('handle typing start events', () {
+    test('add typists in separate narrows', () {
+      prepareModel();
+
+      model.handleTypingEvent(
+        mkTypingEvent(dmNarrow, TypingOp.start, otherUser.userId));
+      check(model).typistsByLocation.deepEquals(
+        {dmNarrow: {otherUser.userId: anyTimer}});
+      checkNotifiedOnce();
+
+      model.handleTypingEvent(
+        mkTypingEvent(groupNarrow, TypingOp.start, thirdUser.userId));
+      check(model).typistsByLocation.deepEquals({
+        dmNarrow: {otherUser.userId: anyTimer},
+        groupNarrow: {thirdUser.userId: anyTimer}});
+      checkNotifiedOnce();
+
+      model.handleTypingEvent(
+        mkTypingEvent(topicNarrow, TypingOp.start, eg.selfUser.userId));
+      check(model).typistsByLocation.deepEquals({
+        dmNarrow: {otherUser.userId: anyTimer},
+        groupNarrow: {thirdUser.userId: anyTimer},
+        topicNarrow: {eg.selfUser.userId: anyTimer}});
+      checkNotifiedOnce();
+    });
+
+    test('add a typist in the same narrow', () {
+      prepareModel();
+
+      model.handleTypingEvent(
+        mkTypingEvent(groupNarrow, TypingOp.start, eg.selfUser.userId));
+      check(model).typistsByLocation.deepEquals({
+        groupNarrow: {eg.selfUser.userId: anyTimer}});
+      checkNotifiedOnce();
+
+      model.handleTypingEvent(
+        mkTypingEvent(groupNarrow, TypingOp.start, thirdUser.userId));
+      check(model).typistsByLocation.deepEquals({
+        groupNarrow: {eg.selfUser.userId: anyTimer, thirdUser.userId: anyTimer}});
+      checkNotifiedOnce();
+    });
+
+    test('sort dm recipients', () {
+      prepareModel();
+
+      final eventUnsorted = TypingEvent(id: 1, op: TypingOp.start, senderId: eg.selfUser.userId,
+        messageType: MessageType.private,
+        recipientIds: [5, 4, 10, 8, 2, 1, eg.selfUser.userId],
+        streamId: null,
+        topic: null);
+      // DmNarrow's constructor expects the recipient IDs to be sorted.
+      check(() => model.handleTypingEvent(eventUnsorted)).returnsNormally();
+    });
+
+  });
+
+  group('handle typing stop events', () {
+    test('remove a typist from an unknown narrow', () => awaitFakeAsync((async) async {
+      prepareModel();
+
+      model.handleTypingEvent(
+        mkTypingEvent(groupNarrow, TypingOp.start, otherUser.userId));
+      checkNotifiedOnce();
+
+      check(model).typistsByLocation.deepEquals({
+        groupNarrow: {otherUser.userId: anyTimer}});
+
+      model.handleTypingEvent(
+        mkTypingEvent(dmNarrow, TypingOp.stop, otherUser.userId));
+      check(model).typistsByLocation.deepEquals({
+        groupNarrow: {otherUser.userId: anyTimer}});
+      checkNotNotified();
+    }));
+
+    test('remove another typist in a narrow', () => awaitFakeAsync((async) async {
```

Comment:
I'm not sure I understand; would the meaning change if this test were called 'remove a typist in a narrow' instead?

Reply:
This is specifically for testing the case when there are multiple typists but we only want to remove one. Renaming this to "remove one of two typists in the same narrow" might be better.